### PR TITLE
Move datasource scopes and actions to access control package

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
In alerting we need to user actions and scopes of data sources and reference of API package causes cycle of dependencies. 
This PR: 
- introduces a new interface `ScopeProvider` and its implementation. This will help encapsulate the control over scope creation in the package.
- moves data source actions and scopes to the datasource package, and updates the usages to use either provider or constants in the package
- moves folder actions and scopes to `dashboard` package and updates usages.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Review by sakjur — APPROVED** [review]
This is not the right place for them but it’s okay for now, it was a mistake (my mistake, to be clear) to ever place anything related to a specific service in access control and not a package for that service.

However, placing them in the very messy api package was not a better solution, so I’m going to improve this based solely on it being a slight improvement of the status quo.

File: pkg/services/accesscontrol/models.go [hunk 1]
```diff
@@ -356,15 +356,30 @@ const (
...
 	// Team scope
 	ScopeTeamsID = Scope("teams", "id", Parameter(":teamId"))
 
 	// Folder scopes
-	ScopeFolderID = Scope("folders", "id", Parameter(":id"))
+	ScopeFoldersAll = GetResourceAllScope(ScopeFoldersRoot)
+	ScopeFolderID   = GetResourceScope(ScopeFoldersRoot, Parameter(":id"))
+
+	// Datasource scopes
+	ScopeDatasourcesAll = Scope(ScopeDatasourcesRoot, "*")
```
Comment (IevaVasiljeva) [inline pkg/services/accesscontrol/models.go:382]: nit: could use `GetResourceAllScope` for this.

File: pkg/api/accesscontrol.go [hunk 2]
```diff
@@ -37,10 +30,9 @@ var (
 	ScopeProvisionersDatasources   = ac.Scope("provisioners", "datasources")
 	ScopeProvisionersNotifications = ac.Scope("provisioners", "notifications")
 
-	ScopeDatasourcesAll = ac.Scope("datasources", "*")
-	ScopeDatasourceID   = ac.Scope("datasources", "id", ac.Parameter(":id"))
-	ScopeDatasourceUID  = ac.Scope("datasources", "uid", ac.Parameter(":uid"))
-	ScopeDatasourceName = ac.Scope("datasources", "name", ac.Parameter(":name"))
+	ScopeDatasourceID   = ac.GetResourceScope(ac.ScopeDatasourcesRoot, ac.Parameter(":id"))
+	ScopeDatasourceUID  = ac.GetResourceScopeUID(ac.ScopeDatasourcesRoot, ac.Parameter(":uid"))
+	ScopeDatasourceName = ac.GetResourceScopeName(ac.ScopeDatasourcesRoot, ac.Parameter(":name"))
```
Comment (IevaVasiljeva) [inline pkg/api/accesscontrol.go:35]: nit: let's also move these data source scopes to `pkg/services/accesscontrol/models.go` to have all data source related actions and scopes in one place.

**Review by IevaVasiljeva — COMMENTED** [review]
Agreed with Emil, in future we will have to rethink where we place scopes and actions, but this works for now :)

**Review by IevaVasiljeva — APPROVED** [review]

**Review by gamab — APPROVED** [review]

File: pkg/api/accesscontrol.go (hunk 2 above)
Comment (yuri-tceretian) [reply pkg/api/accesscontrol.go:35]: The reason I left them close to where they're used is the `Parameter`, which can be different in different use cases. For example, in alerting API it is `:Recipient` not `:id`.

**Review by sakjur — APPROVED** [review]

**Pull request merged**